File: 7.SecondLargest.py
# Find the second largest element using divide and conquer.
import math

def findSecondLargest(arr, l, r):
    # Base case: single element
    if l == r:  # O(1)
        return arr[l], -math.inf  

    # Base case: two elements
    if r == l + 1:  # O(1)
        if arr[l] > arr[r]:
            return arr[l], arr[r]
        else:
            return arr[r], arr[l]

    mid = (l + r) // 2  # O(1)

    # Recursive calls → T(n/2) + T(n/2)
    left = findSecondLargest(arr, l, mid)   
    right = findSecondLargest(arr, mid + 1, r)

    # Merge results → O(1)
    if left[0] > right[0]:
        largest = left[0]
        second_largest = max(left[1], right[0])
    else:
        largest = right[0]
        second_largest = max(right[1], left[0])

    return largest, second_largest

# Driver
arr = [10, 20, 5, 40, 30, 50]
largest, second_largest = findSecondLargest(arr, 0, len(arr) - 1)
print("Largest:", largest, "Second Largest:", second_largest)

"""
Time Complexity:
- Each recursive step = O(1) merge
- Recurrence: T(n) = 2T(n/2) + O(1) → O(n)

Space Complexity:
- Recursion stack = O(log n)
- Extra variables = O(1)
- Total = O(log n)
"""

File: 7.SecondLargest.cpp
// Find the second largest element using divide and conquer.
#include <bits/stdc++.h>
using namespace std;

pair<int, int> findSecondLargest(vector<int>& arr, int l, int r) {
    // Base case: if only one element, no second largest
    if (l == r) return {arr[l], INT_MIN}; // O(1) time, O(1) space

    // Base case: if two elements
    if (r == l + 1) { // O(1)
        if (arr[l] > arr[r]) return {arr[l], arr[r]}; // largest, second largest
        else return {arr[r], arr[l]};
    }

    int mid = (l + r) / 2; // O(1)

    // Recursive calls → T(n/2) + T(n/2)
    pair<int, int> left = findSecondLargest(arr, l, mid);   // O(log n) depth
    pair<int, int> right = findSecondLargest(arr, mid + 1, r);

    int largest, secondLargest;

    // Compare results from left and right → O(1)
    if (left.first > right.first) {
        largest = left.first;
        secondLargest = max(left.second, right.first);
    } else {
        largest = right.first;
        secondLargest = max(right.second, left.first);
    }

    return {largest, secondLargest}; // O(1)
}

int main() {
    vector<int> arr = {10, 20, 5, 40, 30, 50};
    pair<int, int> ans = findSecondLargest(arr, 0, arr.size() - 1);

    cout << "Largest: " << ans.first << ", Second Largest: " << ans.second << endl;
    return 0;
}

/*
Time Complexity (line by line):
- Base cases → O(1)
- Recursive division → T(n) = 2T(n/2) + O(1)
- Master Theorem → O(n)

Space Complexity (line by line):
- Recursion depth = O(log n) (stack space)
- Extra variables = O(1)
- Total SC = O(log n)
*/
